Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,200 @@
+binding and a release callback that is invoked when the device is unregistered.
+
+.. code-block:: c
+
+	struct ancillary_device {
+		struct device dev;
+		const char *match_name;
+		void (*release)(struct ancillary_device *);
+		u32 id;
+	};
+
+The ancillary device is enumerated when it is attached to the bus. The device
+is assigned a unique ID automatically that will be appended to its name. If
+two ancillary_devices both named "foo" are registered onto the bus, they will
+have the device names, "foo.x" and "foo.y", where x and y are unique integers.
+
+Ancillary Device Memory Model and Lifespan
+------------------------------------------
+
+There will be a parent struct, defined in the shared header file, that contains
+the ancillary_device as an static element and a pointer to the shared struct,
+also defined in the shared header.  Both the parent struct and the shared struct
+will be allocated by the ancillary_device's registering driver.  This layout
+allows the ancillary_driver's registering driver to perform a container_of()
+call to go from the pointer to the ancillary_device, that is passed during the
+call to the ancillary_driver's probe function, up to the parent struct, and then
+have access to the shared struct.
+
+The memory for the ancillary_device will be freed only in its release()
+callback flow as defined by its registering driver.
+
+The memory for the shared struct must have a lifespan equal to, or greater than,
+the lifespan of the memory for the ancillary_device.  The ancillary_driver
+should only consider that this shared struct is valid as long as the
+ancillary_device is still registered on the ancillary bus.  It is up to the
+ancillary_device's registering driver to manage (e.g. free or keep available)
+the memory for the shared struct beyond the life of the ancillary_device.
+
+Ancillary Drivers
+=================
+
+Ancillary drivers follow the standard driver model convention, where
+discovery/enumeration is handled by the core, and drivers
+provide probe() and remove() methods. They support power management
+and shutdown notifications using the standard conventions.
+
+.. code-block:: c
+
+	struct ancillary_driver {
+		int (*probe)(struct ancillary_device *);
+		int (*remove)(struct ancillary_device *);
+		void (*shutdown)(struct ancillary_device *);
+		int (*suspend)(struct ancillary_device *, pm_message_t);
+		int (*resume)(struct ancillary_device *);
+		struct device_driver driver;
+		const struct ancillary_device_id *id_table;
+	};
+
+Ancillary drivers register themselves with the bus by calling
+ancillary_register_driver(). The id_table contains the names of ancillary devices
+that a driver can bind with?
+
+Example Usage
+=============
+
+Ancillary devices are created and registered by a subsystem-level core device
+that needs to break up its functionality into smaller fragments. One way to
+extend the scope of a ancillary_device would be to encapsulate it within a
```

Comment:
there are a few occurrences of "a ancillary..." in this document left after automatic substitutions.